### PR TITLE
txoutsbyaddress index

### PR DESCRIPTION
**Pull request by cozz**
Adds new rpc call "gettxoutsbyaddress" as requested in #4007.
Disabled by default, enabled with -txoutsbyaddressindex.

The index is built from the normal utxo on first use, reindex is not required.
For the GUI there is progress shown, on bitcoind you just need to be patient.

The qa-test includes a simple reorganization.
I tested all 4 code parts in main.cpp by commenting them out and check if the qa-test fails at the expected line.
I had to modify the rpc call for this test, as normally wrong outputs in the address index are not exposed to the user, because the rpc call relies on the normal GetCoins call.
The rest of the code has been tested in random ad-hoc testing.
I only tested linux on my machine.

Its all in 1 commit for laziness reasons, but if splitting in multiple commits would help reviewing, I could do that.

txoutsbyaddress

txoutsbyaddress3

### CODE REVIEW

**Comment by gwillen** [comment]
In #bitcoin, the user proserpine- reported the following segfault while running this pullreq:

http://pastebin.com/raw.php?i=D8EUbu4W

He says it happens once every few weeks. I've been looking at it and chatting with @gmaxwell about it, and it seems like it might be related to using 'tx' in CTxMemPool::remove after it's been invalidated.

In particular, it looks to me like tx is always a reference into mapNextTx, and my understanding is that if we reference it after performing _any_ removal from mapNextTx -- which does happen -- we run the risk that our reference will become invalid (e.g. if mapNextTx undergoes a resize due to the removal.)

It looks like the other code in the function carefully avoids touching tx after removing from mapNextTx. (Also, at HEAD the remove function has been rewritten to be non-recursive, and is even more careful in how it uses the value of tx, which is now origTx.)

It looks to me like it might be sufficient (in both the old and new versions of remove) just to move your code to the top of remove(), but I'm just eyeballing that and you should not take my word for it.

**Comment by gwillen** [comment]
Hmm, I take back part of that -- it looks like references into a map _other than_ the one being removed are not invalidated by a removal. So it's no longer totally clear to me that tx is being invalidated and then used. But it still seems like a conceivable explanation for the crash. I am going to keep looking.

**Comment by gwillen** [comment]
Ok, I think I understand now -- I was mistaken about the transaction being in mapNextTx, which only stores pointers to transactions in mapTx. And we're erasing the transaction from mapTx, and then later accessing it. So the new code just needs to be moved above the place where we remove the transaction from mapTx thus invalidating our reference to it.

**Comment by gwillen** [comment]
Oops, that paste expired, here's a better one: 

http://pastebin.com/raw.php?i=s0uqLhtZ

**Comment by paveljanik** [comment]
Can you please rebase this?

**Comment by cozz** [comment]
Rebased and moved the code in CTxMemPool::remove(..), which probably fixes the segfault. 
thanks @gwillen

**Comment by paveljanik** [comment]
Works ok here, no crashes. Progress is displayed correctly when building the index etc.

What I consider strange is `=0` semantic to remove the index.
I'd go with `-txoutsbyaddressindex` to enable working with the index and if it is not found, build it.
And `-deletetxoutsbyaddressindex`. Quite long, but better IMO. Or maybe this one should be the RPC command, not command line argument. What do you think?

Or the other way. Make it all RPC commands: `txoutbyaddress true` and `... false`. And use the index automatically when it is built.

**Comment by Diapolo** [comment]
We have the same behaviour for ``-txindex``. If the flag is present it's enabled and if it's just removed or set to ``-txindex=0``, the index is also removed. IMHO this is expected behaviour.

**Comment by paveljanik** [comment]
@Diapolo Yes, but how GUI users do use this? Via RPC they could.

**Comment by paveljanik** [comment]
This need rebase now.

File: qa/rpc-tests/txoutsbyaddress.sh [hunk 1]
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env bash
+# Copyright (c) 2013-2014 The Bitcoin Core developers
+# Distributed under the MIT/X11 software license, see the accompanying
```
Comment (fanquake) [inline qa/rpc-tests/txoutsbyaddress.sh]: Can you drop the /X11 mentions from this pull before rebasing

**Comment by cozz** [comment]
Rebased, removed /X11.

**Comment by paveljanik** [comment]
What will happen when the initial building of index is running (takes a lot of time here...) and the user pressed Ctrl+C to interrupt? The shutdown window is shown and the user has to wait anyway? This has to be changed IMO...

**Comment by paveljanik** [comment]
Travis build failure on OS X can be ignored...

**Comment by paveljanik** [comment]
Hmm, wanted to try it, but this again needs rebase.

**Comment by paveljanik** [comment]
rpcserver.cpp needs rebase - very simple one.

**Comment by cozz** [comment]
ok, rebased once again.

**Comment by paveljanik** [comment]
tested ACK
More ideas: RPC call `getaddresseswithtxouts`; be able to stop the index creation in the middle.
But anyway, thanks for nice work.

**Comment by paveljanik** [comment]
This needs more review... And github says that even a rebase is needed again.

**Comment by ecdsa** [comment]
it needs to be rebased again...

**Comment by cozz** [comment]
Rebased.

**Comment by ecdsa** [comment]
hi can you rebase this again? thanks

**Comment by cozz** [comment]
Rebased.

**Comment by jgarzik** [comment]
Leaning towards closing this PR.  This garnered some interest, but no ACKs after a long time.  It is not fair to ask @cozz to continually rebase if it's not going in, in the short/medium term.

**Comment by cozz** [comment]
Yes, a decision for 0.12 would be good. Otherwise spending more time on this, just feels like a waste to me.

**Comment by btcdrak** [comment]
I have this PR on my list to test this week.

**Comment by btcdrak** [comment]
Sorry for the delay, I finally tested it yesterday and today on a random sample set of 0 confirms and old txs.

Tested ACK

**Comment by dcousens** [comment]
@btcdrak,  I haven't yet tested this,  but fundamentally what makes this so much larger than [your patch](https://github.com/btcdrak/bitcoin/commit/d48d9c0e9b0cbbca196957d4fa1c5da56431e51e)?

File: src/txdb.cpp [hunk 2]
```diff
@@ -35,17 +36,33 @@ void static BatchWriteCoins(CLevelDBBatch &batch, const uint256 &hash, const CCo
         batch.Write(make_pair(DB_COINS, hash), coins);
 }
 
+void static BatchWriteCoins(CLevelDBBatch &batch, const uint160 &hash, const CCoinsByScript &coins) {
```
Comment (sipa) [inline src/txdb.cpp]: Can you name this function differently? I found it confusing to see it being used by the by-script logic.

File: src/txdb.cpp [hunk 3]
```diff
@@ -35,17 +36,33 @@ void static BatchWriteCoins(CLevelDBBatch &batch, const uint256 &hash, const CCo
...
+    if (coins.IsEmpty())
+        batch.Erase(make_pair('d', hash));
```
Comment (sipa) [inline src/txdb.cpp]: Can you introduce a DB_COINS_BYSCRIPT constant above instead of 'd'?

**Comment by sipa** [comment]
Agree with @jgarzik, it's not fair to keep this lingering for too long.

I like the use case it solves, but I still dislike the degree to which it's entangled with the CCoinsView. I think this feature should be something that is independent, rather than complicating the code necessary for its core function.

Would it be reasonable to implement it as a separate database file, and have it function like the wallet? Something that registers with CValidationInterface, and listens for new transactions being added (and removed) from the blockchain. It could store its own CBlockLocator (like the wallet does, triggered by CValidationInterface's SetBestChain) so it can "rescan" like the wallet does too.

A by-script index for the mempool could be done separately, using an extra index added after #6470's multiindex for the mempool?

**Comment by cozz** [comment]
Well, I am sorry, I am closing this now. As you may have noticed, I am not contributing to this project anymore, because I disagree with basically anything here and I am doing something else now.  If anybody wants to pick this up, feel free to just copy and paste from my code.

Off-topic:
I disagree with most code changes you guys are making. For example main.cpp was much more readable in 0.8 than it is now. Or libconsensus. You guys are very good low-level coders, but you seriously have zero understanding of what encapsulation of problems is. (hint: its not moving code to another file) You are actually making the source code worse in my opinion. 

You havent merged my wallet-pulls, which I consider as important bug-fixes because the wallet-performance is just a bug. Instead you consider moving code around as more important. Long-term goal is to remove GUI and wallet even.

And last but not least, the block size. It is your responsibility to give people the latest in technology. This includes scale. Your job here is to max out the block size.  This is like if your job is to make a 20000 people stadium full and you only let 1000 people in, with the argument that you can never satisfy demand anyway.

You are acting like, if we go higher than 1MB, then suddenly everybody has to fully trust the miners. The threshold to rent a server for a day and verify all blocks to not be affordable to normal people is very much higher than 1MB.

I believe we need different people in the lead position here to be successful. At least gavin should 
be back on the throne with final say on really everything.

So sorry, but long-term I have given up on the project. I am still following a little, so if there are maintanance things for my code, feel free to contact.

**Comment by sipa** [comment]
I'm sorry you feel that way, and I understand the sentiment also partially.

I agree moving code to another file is not proper encapsulation, but some
of the time, moving is a first step, and an easily reviewable one.
Especially the separation of consensus code is not something we want to do
without very high assurances behaviour does not change. I think we are
making progress, but it is admittedly slow.

My only suggestion above here was a more encapsulated approach, by the way,
not a NAK...

Regarding total complexity: I fully agree the code is not "nice". It wasn't
nice when Satoshi disappeared, and it still isn't. Significant parts have
been properly encapsulated since (script execution, addrman, wallet, UTXO
management, network base, UI interaction, limited size data structures),
but at the same time, the code definitely has grown too. I believe that was
necessary: higher performance often means the problem becomes more complex
(for example the introduction of caches, batch processing, precomputed
values, ...). I often very much wish to rewrite everything in a clean way,
but that's just not something that would be reviewable.

About your wallet patches not being merged: I'm aware. But the truth is
that there is little interest in the wallet code, from both users and
developers. We need people who are interested in reviewing and testing to
make changes. That of course leads to the wallet code growing more and more
outdated, reinforcing that circle. One solution to this may be
@jonasschnelli's new wallet project, which aims to add a new expiremental
wallet from scratch, which would be more competitive in features.

Regarding block size: please discuss that on the mailing list. As a Bitcoin
Core maintainer, I believe we should merge any consensus change that
appears uncontroversial to the community.

**Comment by cozz** [comment]
I didnt expect such an honest response, thanks, that gives you a lot of respect.
It doesnt make me change my mind about this project though.

**Comment by rawmaterials** [comment]
how quick does this returns data ? 
has anybody tried it ?

**Comment by shamoons** [comment]
Still open?

**Pull request closed without merging**